db/schema_tables.cc

```diff
@@ -2684,6 +2684,20 @@ static index_metadata create_index_from_index_row(const query::result_set_row& r
     return index_metadata{index_name, options, kind, is_local};
 }
 
+// FIXME: remove this fix after a full release cycle, after which it's certain
+// that all secondary index schemas will be fixed during an upgrade path
+// Ref: #7513
+static void maybe_fix_secondary_index_computed_column(column_definition& cdef, sstring_view ks_name, sstring_view cf_name) {
+    if (to_sstring_view(cdef.name()) == "idx_token"
+            && cdef.kind == column_kind::clustering_key
+            && cdef.id == 0
+            && !cdef.is_computed()
+            && boost::algorithm::ends_with(cf_name, "_index")) {
```

Comment:
Can we add more conditions, to reduce the probabilty that a user table that happens to be named something_index gets misclassified here?

Reply:
The current conditions are:
1. It's a materialized view (this function is only used for views)
2. It has a `idx_token` column ...
3. ... as a first clustering key ...
4. and this view's name ends with `_index`

I think that `2` is the least likely of these, but I wish we named it `__scylla_idx_token` back in the days. This point in code is quite elegant to fix the index, but its one disadvantage is that we can't assume that other tables are loaded yet.

Ideally, we should also check that this table actually corresponds to a secondary index, but that can only be done later. The problem with checking later is that the index is already created, so that involves triggering a schema change announcement locally, and I worry that it's prone to races, because there's a window in which the index schema is incorrect, and also I'm not sure how well such internal schema change requests coexist with waiting for schema agreement, etc.

Reply:
@psarna 

> Ideally, we should also check that this table actually corresponds to a secondary index, but that can only be done later.

I also encountered it as a limitation when working on some other fix, I am starting to think, maybe we should only load materialized views after all "main" tables have been loaded. This can actually solve a lot of races and allow us to have the invariant " If a view exists it's base table exists".

There is still a small nit, I remember seeing this whole piece of code whose purpose is to determine this idx_token name - in the 0.000001% chance that someone had this column as one of it's PK columns :slightly_smiling_face: , in that case I believe there would still be a crash right?
Maybe we need to go for a pattern (_*idx_token)  and a position in the PK (meaning the first clustering key), unless I misunderstood this token purpose, it must be the first clustering key, no?

Reply:
Oh, right, purely theoretically, the name can be `idx_token_N`, with `N` being a natural number.

Reply:
I think that the name idx_token is ugly enough. I missed that it was a precondition.

We use the `$` to mark internal identifiers. e.g. `cdc$blah`. So the internal tables could have been named `blah$index`.